// shazam-api/dist/api.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Shazam = exports.Endpoint = void 0;
const algorithm_1 = require("./algorithm");
const node_fetch_1 = __importDefault(require("node-fetch"));
const TIME_ZONE = "Europe/Paris";
function uuidv4() {
    return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function (c) {
        var r = Math.random() * 16 | 0, v = c == 'x' ? r : (r & 0x3 | 0x8);
        return v.toString(16);
    }).toUpperCase();
}
class Endpoint {
    constructor(timezone) {
        this.timezone = timezone;
    }
    ;
    url() {
        return `${Endpoint.SCHEME}://${Endpoint.HOSTNAME}/discovery/v5/en/US/iphone/-/tag/${uuidv4()}/${uuidv4()}`;
    }
    params() {
        return {
            'sync': 'true',
            'webv3': 'true',
            'sampling': 'true',
            'connected': '',
            'shazamapiversion': 'v3',
            'sharehub': 'true',
            'hubv5minorversion': 'v5.1',
            'hidelb': 'true',
            'video': 'v3'
        };
    }
    headers() {
        return {
            "X-Shazam-Platform": "IPHONE",
            "X-Shazam-AppVersion": "14.1.0",
            "Accept": "*/*",
            "Content-Type": "application/json",
            "Accept-Encoding": "gzip, deflate",
            "Accept-Language": "en",
            "User-Agent": "Shazam/3685 CFNetwork/1197 Darwin/20.0.0"
        };
    }
    async sendRecognizeRequest(url, body) {
        var _a;
        const fetch = (_a = global.fetch) !== null && _a !== void 0 ? _a : node_fetch_1.default;
        return await (await fetch(url, { body, headers: this.headers(), method: "POST" })).json();
    }
    async formatAndSendRecognizeRequest(signature) {
        let data = {
            'timezone': this.timezone,
            'signature': {
                'uri': signature.encodeToUri(),
                'samplems': Math.round(signature.numberSamples / signature.sampleRateHz * 1000)
            },
            'timestamp': new Date().getTime(),
            'context': {},
            'geolocation': {}
        };
        const url = new URL(this.url());
        Object.entries(this.params()).forEach(([a, b]) => url.searchParams.append(a, b));
        let response = await this.sendRecognizeRequest(url.toString(), JSON.stringify(data));
        if (response.matches.length === 0)
            return null;
        return response;
    }
}
exports.Endpoint = Endpoint;
Endpoint.SCHEME = "https";
Endpoint.HOSTNAME = "amp.shazam.com";
class Shazam {
    constructor(timeZone) {
        this.endpoint = new Endpoint(timeZone !== null && timeZone !== void 0 ? timeZone : TIME_ZONE);
    }
    async recognizeSong(samples, callback) {
        var _a, _b;
        let response = await this.fullRecognizeSong(samples, callback);
        if (!response)
            return null;
        const trackData = response.track, mainSection = trackData.sections.find((e) => e.type === "SONG");
        const title = trackData.title, artist = trackData.subtitle, album = (_a = mainSection.metadata.find(e => e.title === "Album")) === null || _a === void 0 ? void 0 : _a.text, year = (_b = mainSection.metadata.find(e => e.title === "Released")) === null || _b === void 0 ? void 0 : _b.text;
        return { title, artist, album, year };
    }
    async fullRecognizeSong(samples, callback) {
        callback === null || callback === void 0 ? void 0 : callback("generating");
        let generator = this.createSignatureGenerator(samples);
        while (true) {
            callback === null || callback === void 0 ? void 0 : callback("generating");
            const signature = generator.getNextSignature();
            if (!signature) {
                break;
            }
            callback === null || callback === void 0 ? void 0 : callback("transmitting");
            let results = await this.endpoint.formatAndSendRecognizeRequest(signature);
            if (results !== null)
                return results;
        }
        return null;
    }
    createSignatureGenerator(samples) {
        let signatureGenerator = new algorithm_1.SignatureGenerator();
        signatureGenerator.feedInput(samples);
        return signatureGenerator;
    }
}
exports.Shazam = Shazam;
Shazam.MAX_TIME_SCEONDS = 8;
